src/libcore/str/pattern.rs

```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+    // left to right. If v matches, we try to match u by scanning right to left.
+    // How far we can jump when we encounter a mismatch is all based on the fact
+    // that (u, v) is a critical factorization for the needle.
+    #[inline]
+    fn next(&mut self, haystack: &[u8], needle: &[u8], long_period: bool)
+            -> SearchStep {
+        // `next()` uses `self.position` as its cursor
+        let start = self.position;
+        'search: loop {
+            // Check that we have room to search in
+            if self.position + needle.len() > haystack.len() {
+                if start == haystack.len() {
+                    return SearchStep::Done
+                } else {
+                    self.position = haystack.len();
+                }
+            }
+            if start != self.position {
+                return SearchStep::Reject(start, self.position);
+            }
+
+            // Quickly skip by large portions unrelated to our substring
+            if (self.byteset >>
+                    ((haystack[self.position + needle.len() - 1] & 0x3f)
+                     as usize)) & 1 == 0 {
+                self.position += needle.len();
+                if !long_period {
+                    self.memory = 0;
+                }
+                continue 'search;
+            }
+
+            // See if the right part of the needle matches
+            let start = if long_period { self.crit_pos }
+                        else { cmp::max(self.crit_pos, self.memory) };
+            for i in start..needle.len() {
+                if needle[i] != haystack[self.position + i] {
+                    self.position += i - self.crit_pos + 1;
+                    if !long_period {
+                        self.memory = 0;
+                    }
+                    continue 'search;
+                }
+            }
+
+            // See if the left part of the needle matches
+            let start = if long_period { 0 } else { self.memory };
+            for i in (start..self.crit_pos).rev() {
+                if needle[i] != haystack[self.position + i] {
+                    self.position += self.period;
+                    if !long_period {
+                        self.memory = needle.len() - self.period;
+                    }
+                    continue 'search;
+                }
+            }
+
+            // We have found a match!
+            let match_pos = self.position;
+            self.position += needle.len(); // add self.period for all matches
+            if !long_period {
+                self.memory = 0; // set to needle.len() - self.period for all matches
+            }
+
+            return SearchStep::Match(match_pos, match_pos + needle.len());
+        }
+    }
+
+    // Follows the ideas in `next()`.
+    //
+    // Note: No documentation has been used for the algorithm in `next_back()`,
+    // simply trying to adapt forward search to backward search with extensive
+    // testing.
```

Comment:
Quick proof that this approach should work:

All the definitions are completely symmetrical, with `period(x) = period(reverse(x))` and `local_period(u, v) = local_period(reverse(v), reverse(u))`, so if `(u, v)` is a critical factorization, so is `(reverse(v), reverse(u))`. Similarly, the "period" stored in `self.period` is the real period if `long_period` is `false`, and so is still valid for a reversed needle, and if `long_period` is true, all the algorithm requires is that `self.period` is less than or equal to the real period, which must be true for the forward case anyway, so the value stored in `self.period` is still valid.

To search in reverse through the haystack, we search forward through a reversed haystack with a reversed needle, and the above paragraph shows that the precomputed parameters can be left alone, which is exactly what this code implements.